File: adapters/go/sdk/examples/modbus_tcp_sensor/decoder.go
package main

import (
	"encoding/binary"
	"fmt"
	"math"

	"github.com/e7217/edg/adapters/go/sdk"
)

// RegisterSpec mirrors one row of the mapping YAML and the Python
// RegisterSpec dataclass. Field tags map to the YAML schema.
type RegisterSpec struct {
	Name      string  `yaml:"name"`
	Function  string  `yaml:"function"`  // "holding" | "input"
	Address   uint16  `yaml:"address"`
	Type      string  `yaml:"type"`      // uint16 | int16 | uint32 | int32 | float32
	WordOrder string  `yaml:"word_order"`
	Scale     float64 `yaml:"scale"`
	Unit      string  `yaml:"unit"`
}

// WordCount returns how many 16-bit registers this spec consumes.
func (r RegisterSpec) WordCount() (uint16, error) {
	switch r.Type {
	case "uint16", "int16":
		return 1, nil
	case "uint32", "int32", "float32":
		return 2, nil
	default:
		return 0, fmt.Errorf("unsupported type: %s", r.Type)
	}
}

// DecodeRegister converts raw 16-bit register words into a TagValue.
// The Quality is always GOOD; the adapter is responsible for downgrading
// quality on read failure.
func DecodeRegister(words []uint16, spec RegisterSpec) (sdk.TagValue, error) {
	scale := spec.Scale
	if scale == 0 {
		scale = 1.0
	}

	var value float64
	switch spec.Type {
	case "uint16":
		if len(words) != 1 {
			return sdk.TagValue{}, fmt.Errorf("uint16 requires 1 word, got %d", len(words))
		}
		value = float64(words[0])

	case "int16":
		if len(words) != 1 {
			return sdk.TagValue{}, fmt.Errorf("int16 requires 1 word, got %d", len(words))
		}
		value = float64(int16(words[0]))

	case "uint32", "int32", "float32":
		if len(words) != 2 {
			return sdk.TagValue{}, fmt.Errorf("%s requires 2 words, got %d", spec.Type, len(words))
		}
		buf, err := toBigEndianBytes(words, spec.WordOrder)
		if err != nil {
			return sdk.TagValue{}, err
		}
		switch spec.Type {
		case "uint32":
			value = float64(binary.BigEndian.Uint32(buf))
		case "int32":
			value = float64(int32(binary.BigEndian.Uint32(buf)))
		case "float32":
			value = float64(math.Float32frombits(binary.BigEndian.Uint32(buf)))
		}

	default:
		return sdk.TagValue{}, fmt.Errorf("unsupported type: %s", spec.Type)
	}

	n := value * scale
	return sdk.TagValue{
		Name:    spec.Name,
		Quality: sdk.QualityGood,
		Number:  &n,
		Unit:    spec.Unit,
	}, nil
}

// toBigEndianBytes reassembles two raw register words into a canonical
// big-endian 4-byte buffer, undoing any vendor-specific byte/word swap.
//
// Word-order convention (vendor manual style):
//
//	ABCD: bytes [A B C D] -> word0=AB, word1=CD (big-endian, normal)
//	CDAB: word swap -> word0=CD, word1=AB
//	BADC: byte swap inside each word -> word0=BA, word1=DC
//	DCBA: both swapped -> word0=DC, word1=BA
func toBigEndianBytes(words []uint16, order string) ([]byte, error) {
	hi, lo := words[0], words[1]
	a, b := byte(hi>>8), byte(hi&0xFF)
	c, d := byte(lo>>8), byte(lo&0xFF)

	switch order {
	case "ABCD", "":
		return []byte{a, b, c, d}, nil
	case "CDAB":
		return []byte{c, d, a, b}, nil
	case "BADC":
		return []byte{b, a, d, c}, nil
	case "DCBA":
		return []byte{d, c, b, a}, nil
	default:
		return nil, fmt.Errorf("unsupported word_order: %s (expected one of ABCD, CDAB, BADC, DCBA)", order)
	}
}

File: adapters/python/examples/modbus_tcp/decoder.py
"""Pure register decoder.

Splitting this out from the adapter lets us cover every type x word_order
combination with a fast unit-test matrix that is shared with the Go
implementation via ``testdata/decoder_matrix.yaml``.

Word-order naming follows the convention vendors print in PLC manuals:
``ABCD`` (normal big-endian), ``CDAB`` (word swap), ``BADC`` (byte swap
inside each word), ``DCBA`` (both swapped).
"""

from __future__ import annotations

import struct
from dataclasses import dataclass

from sdk import TagValue


_SCALAR_16 = {"uint16", "int16"}
_SCALAR_32 = {"uint32", "int32", "float32"}
_WORD_ORDERS = ("ABCD", "CDAB", "BADC", "DCBA")


@dataclass
class RegisterSpec:
    """One row in the mapping YAML."""

    name: str
    function: str           # "holding" | "input"
    address: int
    type: str               # "uint16" | "int16" | "uint32" | "int32" | "float32"
    word_order: str = "ABCD"
    scale: float = 1.0
    unit: str = ""

    @property
    def word_count(self) -> int:
        if self.type in _SCALAR_16:
            return 1
        if self.type in _SCALAR_32:
            return 2
        raise ValueError(f"unsupported type: {self.type}")


def decode_register(words: list[int], spec: RegisterSpec) -> TagValue:
    """Decode raw 16-bit register words into a TagValue.

    Returns a TagValue with quality=GOOD. The adapter is responsible for
    downgrading quality on read errors before publish.
    """
    if spec.type in _SCALAR_16:
        if len(words) != 1:
            raise ValueError(f"{spec.type} requires 1 word, got {len(words)}")
        raw = words[0] & 0xFFFF
        if spec.type == "int16":
            raw = _sign_extend(raw, 16)
        value = float(raw)

    elif spec.type in _SCALAR_32:
        if len(words) != 2:
            raise ValueError(f"{spec.type} requires 2 words, got {len(words)}")
        if spec.word_order not in _WORD_ORDERS:
            raise ValueError(
                f"unsupported word_order: {spec.word_order} "
                f"(expected one of {_WORD_ORDERS})"
            )
        raw_bytes = _to_big_endian_bytes(words, spec.word_order)
        if spec.type == "uint32":
            value = float(int.from_bytes(raw_bytes, "big", signed=False))
        elif spec.type == "int32":
            value = float(int.from_bytes(raw_bytes, "big", signed=True))
        else:  # float32
            value = struct.unpack(">f", raw_bytes)[0]

    else:
        raise ValueError(f"unsupported type: {spec.type}")

    return TagValue(
        name=spec.name,
        quality="GOOD",
        number=value * spec.scale,
        unit=spec.unit,
    )


def _sign_extend(value: int, bits: int) -> int:
    sign_bit = 1 << (bits - 1)
    return (value & (sign_bit - 1)) - (value & sign_bit)


def _to_big_endian_bytes(words: list[int], word_order: str) -> bytes:
    """Reassemble 4 bytes [A B C D] in canonical big-endian order.

    Given two raw register words exactly as returned by the device, swap
    bytes/words so the result represents the logical big-endian value.
    """
    hi, lo = words[0] & 0xFFFF, words[1] & 0xFFFF
    a, b = (hi >> 8) & 0xFF, hi & 0xFF
    c, d = (lo >> 8) & 0xFF, lo & 0xFF

    layout = {
        "ABCD": (a, b, c, d),
        "CDAB": (c, d, a, b),
        "BADC": (b, a, d, c),
        "DCBA": (d, c, b, a),
    }[word_order]
    return bytes(layout)
